plugins/hypervisors/kvm/src/main/java/com/cloud/hypervisor/kvm/resource/KVMHAMonitor.java

```diff
@@ -35,121 +35,221 @@
-                    storage = conn.storagePoolLookupByUUIDString(uuid);
-                    if (storage == null || storage.getInfo().state != StoragePoolState.VIR_STORAGE_POOL_RUNNING) {
-                        if (storage == null) {
-                            s_logger.debug(String.format("Libvirt storage pool [%s] not found, removing from HA list.", uuid));
-                        } else {
-                            s_logger.debug(String.format("Libvirt storage pool [%s] found, but not running, removing from HA list.", uuid));
+        if(nfsstoragePool != null && !nfsstoragePool.isEmpty()) {
+            synchronized (nfsstoragePool) {
+                Set<String> removedPools = new HashSet<>();
+                for (String uuid : nfsstoragePool.keySet()) {
+                    NfsStoragePool primaryStoragePool = nfsstoragePool.get(uuid);
+                    StoragePool storage;
+                    try {
+                        Connect conn = LibvirtConnection.getConnection();
+                        storage = conn.storagePoolLookupByUUIDString(uuid);
+                        if (storage == null || storage.getInfo().state != StoragePoolState.VIR_STORAGE_POOL_RUNNING) {
+                            if (storage == null) {
+                                s_logger.debug(String.format("Libvirt storage pool [%s] not found, removing from HA list.", uuid));
+                            } else {
+                                s_logger.debug(String.format("Libvirt storage pool [%s] found, but not running, removing from HA list.", uuid));
+                            }
+
+                            removedPools.add(uuid);
+                            continue;
+                        }
+
+                        s_logger.debug(String.format("Found NFS storage pool [%s] in libvirt, continuing.", uuid));
+
+                    } catch (LibvirtException e) {
+                        s_logger.debug(String.format("Failed to lookup libvirt storage pool [%s].", uuid), e);
+
+                        if (e.toString().contains("pool not found")) {
+                            s_logger.debug(String.format("Removing pool [%s] from HA monitor since it was deleted.", uuid));
+                            removedPools.add(uuid);
+                            continue;
                         }
 
-                        removedPools.add(uuid);
-                        continue;
                     }
 
-                    s_logger.debug(String.format("Found NFS storage pool [%s] in libvirt, continuing.", uuid));
+                    String result = null;
+                    for (int i = 1; i <= _heartBeatUpdateMaxTries; i++) {
+                        Script cmd = createHeartBeatCommand(primaryStoragePool, hostPrivateIp, true);
+                        result = cmd.execute();
+
+                        s_logger.debug(String.format("The command (%s), to the pool [%s], has the result [%s].", cmd.toString(), uuid, result));
```

Comment:
```suggestion
                        s_logger.debug(String.format("The command [%s], to the pool [%s], had the result [%s].", cmd.toString(), uuid, result));
```

Reply:
@GutoVeronezi     I modified the code as you requested.